src/dp.rs

```diff
@@ -102,6 +106,21 @@ impl ZCdpBudget {
 
 impl DifferentialPrivacyBudget for ZCdpBudget {}
 
+/// Pure differential privacy budget. (&epsilon;-DP or (&epsilon;, 0)-DP)
```

Comment:
What you have right now isn't pure, though maybe it is close enough as to be statistically indistinguishable.

The conversion of the output (which is a signed bignum!) to the field includes a modulus, which means that the infinitely long tails on the distribution are wrapped continuously.  In effect, you are creating a new truncated distribution that is close to a discrete Laplace.

As a result of the large size of the field, $\delta$ is probably close enough to zero that it won't matter.  That said, if you want certainty about your $\delta$, maybe a better approach is to truncate the distribution so that you can make a firmer commitment regarding $\delta$ (there are a number of papers looking at truncated distributions that you could draw from).

Reply:
I think this still achieves pure DP, due to the reasoning in the comment at the top of `add_iid_noise_to_field_vec()`. I can take a crack at writing a more traditional proof for this mechanism. I expect it should go through okay, because the PDF of the wrapped discrete Laplace distribution is equal to an infinite sum of PDFs of discrete Laplace distributions with offset means. Thus, it should be possible to pair off terms and show they are bounded above correctly.

One key difference between this mechanism and a mechanism using a traditional truncated noise distribution is that the support of the query output is always the same (zero to p-1) and doesn't shift when the true value of the function shifts.

Reply:
I'm not convinced about that.  Consider a small prime field where $p$ is something small, like 31.  In that field, your value might be 10, with a discrete Laplace distribution that includes values below -10 or above 21.  Those will wrap, such that -10 contributes the same value as +21 (for any true value).

I've not seen an argument that supports the view that the probability density function over a limited domain like that is the same as an unbounded discrete Laplace.  Maybe it's OK.  Maybe it's close enough to OK that it doesn't matter when $p$ is very large, like it will be here.  But "pure" means perfect and that's a strong claim to make.

Reply:
This is a bit subtle, so let me try to clarify.

Differential Privacy satisfies the Post-Processing property  (or Data Processing Inequality). This means that if a mechanism is DP, you can apply any function to the output and that is still DP. Thus if the mechanism that adds Laplace noise over integers is DP, we can apply the transformation $T_q$ that maps an integer z to z mod q, that still preserves the privacy properties. This argument applies to computing the noisy noisy count over integers, and then applying the transform $T_q$, and so we are ok as long as we sample from Laplace over integers and apply $T_q$ to the noisy count. Morally, the reason this works is that the probability of any output $y$ is now the sum of probabilities of outputs $y, y+q, y-q, y+2q, y-2q,....$. Since each of those probabilities changes by at most a multiplicative $e^{\epsilon}$ between neighboring databases, their sum also can only change by a multiplicative $e^{\epsilon}$.

Note that this is different from truncating the noise itself, which would indeed need some delta adjustment, and your intuition above is indeed correct for that variant.

Reply:
I just pushed a proof that the mechanism satisfies pure differential privacy, starting from the formulas. Here's a [rendered PDF](https://github.com/user-attachments/files/16072470/Pure.DP.Mechanism.pdf) for convenience; I authored this in LyX, and put the source file in the documentation directory.

Reply:
Thanks both, the intuition and proof both helped, though I did not go through in detail, the inequality is the key piece I was looking for.

Question: is that the last thing holding the merge here?  I was not expecting to be the one holding this up.  We are very interested in having this land.

Reply:
I think this and the neighboring dataset definition thread are sufficiently resolved, but I'd like to give @cjpatton a chance to review before I merge.

Reply:
Thanks @kunal-talwar for giving us some intuition. I've gone the proof. There was a couple of steps I didn't see, but I'm confident David will clarify.